src/logic/gameReducer.js

```diff
@@ -86,6 +86,177 @@ function getConnectedPieceIDs({ pieces, gridSize, draggedPieceID }) {
   return Array.from(touchingIDs);
 }
 
+function dragStart({
+  currentGameState,
+  predicate,
+  pointerID,
+  pointer,
+  pointerOffset,
+  isShifting,
+}) {
+  if (currentGameState.dragState !== undefined) {
+    return currentGameState;
+  }
+
+  // Find which pieces are selected, which are not, and the top left of the group (in board squares).
+  let targets = [];
+  let nontargets = [];
+  let groupBoardTop = currentGameState.gridSize;
+  let groupBoardLeft = currentGameState.gridSize;
+  const poolPieces = currentGameState.pieces.filter(
+    (piece) => piece.poolIndex >= 0
+  );
+  let poolIndex = poolPieces.length;
+  for (const piece of currentGameState.pieces) {
+    if (predicate(piece)) {
+      targets.push(piece);
+      if (groupBoardTop !== undefined) {
+        if (piece.boardTop !== undefined) {
+          groupBoardTop = Math.min(groupBoardTop, piece.boardTop);
+          groupBoardLeft = Math.min(groupBoardLeft, piece.boardLeft);
+        } else {
+          groupBoardTop = undefined;
+          groupBoardLeft = undefined;
+          if (piece.poolIndex !== undefined) {
+            poolIndex = Math.min(poolIndex, piece.poolIndex);
+          }
+        }
+      }
+    } else {
+      nontargets.push(piece);
+    }
+  }
+  if (targets.length === 0) {
+    return currentGameState;
+  }
+
+  // Find the top left of the group in client coordinates, to get pointerOffset.
+  if (pointerOffset === undefined) {
+    const rectangles = targets.flatMap((piece) => {
+      const element = document.getElementById(`piece-${piece.id}`);
+      if (!element) {
+        console.warn(
+          `dragStart: element for piece ${piece.id} not found in DOM`
+        );
+        return [];
+      }
+      return [element.getBoundingClientRect()];
+    });
+    if (rectangles.length === 0) {
+      return currentGameState;
+    }
+    const groupTop = Math.min(...rectangles.map((rect) => rect.top));
+    const groupLeft = Math.min(...rectangles.map((rect) => rect.left));
+    pointerOffset = {
+      x: pointer.x - groupLeft,
+      y: pointer.y - groupTop,
+    };
+  }
+
+  currentGameState = {
+    ...currentGameState,
+    pieces: nontargets.concat(
+      targets.map((piece) => ({
+        ...piece,
+        boardTop: undefined,
+        boardLeft: undefined,
+        poolIndex: undefined,
+        groupTop:
+          groupBoardTop === undefined ? 0 : piece.boardTop - groupBoardTop,
+        groupLeft:
+          groupBoardLeft === undefined ? 0 : piece.boardLeft - groupBoardLeft,
+      }))
+    ),
+    dragCount: currentGameState.dragCount + 1,
+    dragState: {
+      pieceIDs: targets.map((piece) => piece.id),
+      isShifting,
+      dragHasMoved: false,
+      pointerID,
+      pointerStart: pointer,
+      pointer,
+      pointerOffset,
+      destination:
+        groupBoardTop !== undefined
+          ? { where: "board", top: groupBoardTop, left: groupBoardLeft }
```

Comment:
IIUC there is `dragState.destination.where`, which can be `"board"` or `"pool"`, and there is also a `where` input to the `Piece` component, which can be `"board"` or `"pool"` or `"drag"`, correct?

Reply:
Yes, that's right.